---
title: "node_genome_mapping"
output: html_document
date: "2024-11-06"
---

```{r}
library(readr)
library(tidyverse)
library(stringr)
blast_result <- read_table("workflow/out/blast/sample_1_0_02_blast_results.out", guess_max = 40000)

genome_seq_ids <- read_delim("workflow/out/blast/sample_1_reference_metagenome_ids.txt", delim = ";", escape_double = FALSE, trim_ws = TRUE, guess_max = 200)

tax_profile <- read_csv("workflow/out/blast/taxonomic_profile_1.csv")
```

```{r}
tax_table <- tax_profile %>%
  rowwise() %>%
  mutate(kingdom = strsplit(TAXPATHSN, "\\|")[[1]][1], 
         phylum = strsplit(TAXPATHSN, "\\|")[[1]][2],
         class = strsplit(TAXPATHSN, "\\|")[[1]][3],
         order = strsplit(TAXPATHSN, "\\|")[[1]][4],
         family = strsplit(TAXPATHSN, "\\|")[[1]][5],
         genus = strsplit(TAXPATHSN, "\\|")[[1]][6],
         species = strsplit(TAXPATHSN, "\\|")[[1]][7],
         strain = strsplit(TAXPATHSN, "\\|")[[1]][8])
  # mutate(species_stripped = paste(strsplit(species, " ")[[1]][1:2], collapse = " "))


genome_node_clusters <- blast_result %>%
  rowwise %>%
  mutate(node = strsplit(qseqid, ":")[[1]][1]) %>%
  left_join(., genome_seq_ids, by = c("sseqid" = "seqid")) %>%
  select(node, genome, info) %>%
  distinct() %>%
  mutate(species = case_when(genome == "Erysipelotrichaceae bacterium I46" ~ genome,
                             genome == "Actinomyces sp. oral taxon 414 strain F0588" ~ "Actinomyces sp. oral taxon 414",
                             genome == "Coriobacteriaceae bacterium 68-1-3" ~ genome,
                             !str_detect(genome, "\\bsp\\.") ~ 
                               paste(strsplit(genome, " ")[[1]][1:2], collapse = " "),
                             str_detect(genome, "\\bsp\\.") ~ 
                               paste(strsplit(genome, " ")[[1]][1:3], collapse = " "))) %>%
  left_join(., tax_table, by = c("species"))

genome_node_clusters[genome_node_clusters == ""] <- NA

genus_node_clusters <- genome_node_clusters %>%
  rowwise() %>%
  mutate(genus = ifelse(is.na(genus),
                        strsplit(genome, " ")[[1]][1], 
                        genus)) %>%
  select(node, genus, Genera) %>%
  distinct() %>%
  group_by(genus) %>% 
  mutate(Genera = case_when(is.na(Genera) ~ first(Genera[!is.na(Genera)]), 
                            TRUE ~ Genera)) %>%
  ungroup()

genus_node_clusters[genus_node_clusters == ""] <- NA

# manually adding missing values for Genera and genus heh
genus_node_clusters$Genera <- as.character(genus_node_clusters$Genera)
genus_node_clusters$Genera[genus_node_clusters$genus == "Acetivibrio"] <- "35829"
genus_node_clusters$Genera[genus_node_clusters$genus == "Anaerotignum"] <- "2039240"
genus_node_clusters$Genera[genus_node_clusters$genus == "Peptoclostridium"] <- "1481960"


genus_node_clusters$genus[genus_node_clusters$Genera == "128827"] <- "Erysipelotrichaceae"
genus_node_clusters$genus[genus_node_clusters$Genera == "39491"] <- "Agathobacter"
genus_node_clusters$genus[genus_node_clusters$Genera == "2815775"] <- "Berryella"

# determine how we should group clusters by on a genus level -- by genus or genera
nrow(subset(genus_node_clusters, is.na(genus)))
nrow(subset(genus_node_clusters, is.na(Genera)))
# way more rows have values for Genera so I guess we'll use this ugh 


```
